Add Expression constraint to handle general (nonlinear) constraints …

RussTedrake opened the pull request:
…specified by symbolic Expressions

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8338)
<!-- Reviewable:end -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[solvers/constraint.h, line 785 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7PgO_0zUpa994pRTiy:-L7PgO_0zUpa994pRTiz:b-3y1n9g) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/constraint.h#L785)):*
> ```objc
> };
> 
> class ExpressionConstraint : public Constraint {
> ```

FYI, missing doxygen block.

---

*[solvers/mathematical_program.h, line 1046 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7PmSCGgAtzO3AMxdV0:-L7PmSCGgAtzO3AMxdV1:b9xugge) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1046)):*
> ```objc
>    */
>   Binding<Constraint> AddConstraint(const symbolic::Expression& e,
>                                                 double lb, double ub);
> ```

BTW, please check the indentation.

---

*[solvers/mathematical_program.h, line 1097 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7Ppm7Cu--9r4_B4OWB:-L7Ppm7Cu--9r4_B4OWC:b-qmw5ng) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1097)):*
> ```objc
>    *   Eigen::Vector2d b;
>    *   ... // set up A and b
>    *   prog.AddLinearConstraint((A * x).array() <= b.array());
> ```

BTW, `prog.AddLinearConstraint` -> `prog.AddConstraint`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-:-L7PsZHKY5yCFvzn4kpQ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
+@jwnimmer-tri for platform-review, please.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-:-L7PujhMAPXEPtMVRBVE:b4fwciw)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[solvers/evaluator_base.h, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7PwCQ5qsxfUCYBWF5i:-L7PwCQ5qsxfUCYBWF5j:b-jfzvrd) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/evaluator_base.h#L131)):*
> ```objc
>   void set_num_outputs(int num_outputs) { num_outputs_ = num_outputs; }
> 
>   // Setter for the number of variables.
> ```

Could you add more documentation here? None of our other constraints allow to set the number of variables, and we always suppose that the number of variables are constant, after the construction. I feel this setter can open a loop hole in our code.

Actually I think we can avoid to set the number of variables, by making the constructor of `ExpressionConstraint` to be protected, and use a factory method to construct `ExpressionConstraint`? Something like
```C++
class ExpressionConstraint : public Constraint {
 public:
  static std::unique_ptr<ExpressionConstraint> Make(const Eigen::Ref<const VectorX<symbolic::Expression>>& v,
    const Eigen::Ref<const Eigen::VectorXd>& lb,
    const Eigen::Ref<const Eigen::VectorXd>& ub) {
  for (int i = 0; i < expressions_.size(); ++i) {
    internal::ExtractAndAppendVariablesFromExpression(expressions_(i), &vars_,
                                                      &map_var_to_index_);
  }
 return std::make_unique<ExpressionConstraint>(v, vars_.size(), lb, ub);
}
 protected:
  ExpressionConstraint(onst Eigen::Ref<const VectorX<symbolic::Expression>>& v,
    int num_vars,
    const Eigen::Ref<const Eigen::VectorXd>& lb,
    const Eigen::Ref<const Eigen::VectorXd>& ub)
    : Constraint(v.rows(), num_vars, lb, ub) {
}
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 16 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[solvers/constraint.cc, line 184 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7QTJwA2IZVuk2bBKJ9:-L7QTJwA2IZVuk2bBKJA:b1dnljw) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/constraint.cc#L184)):*
> ```cpp
>   }
> 
>   this->set_num_vars(vars_.size());
> ```

BTW `bazel test //solvers:constraint_test` does not fail when I comment out this line, though `//solvers:mathematical_program_test` does.  Do we want to have direct unit coverage here, or is indirect good enough?

---

*[solvers/evaluator_base.h, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7PwCQ5qsxfUCYBWF5i:-L7QHA1oPxd3A91g4hNk:b-2yv10n) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/evaluator_base.h#L131)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Could you add more documentation here? None of our other constraints allow to set the number of variables, and we always suppose that the number of variables are constant, after the construction. I feel this setter can open a loop hole in our code.

Actually I think we can avoid to set the number of variables, by making the constructor of `ExpressionConstraint` to be protected, and use a factory method to construct `ExpressionConstraint`? Something like
```C++
class ExpressionConstraint : public Constraint {
 public:
  static std::unique_ptr<ExpressionConstraint> Make(const Eigen::Ref<const VectorX<symbolic::Expression>>& v,
    const Eigen::Ref<const Eigen::VectorXd>& lb,
    const Eigen::Ref<const Eigen::VectorXd>& ub) {
  for (int i = 0; i < expressions_.size(); ++i) {
    internal::ExtractAndAppendVariablesFromExpression(expressions_(i), &vars_,
                                                      &map_var_to_index_);
  }
 return std::make_unique<ExpressionConstraint>(v, vars_.size(), lb, ub);
}
 protected:
  ExpressionConstraint(onst Eigen::Ref<const VectorX<symbolic::Expression>>& v,
    int num_vars,
    const Eigen::Ref<const Eigen::VectorXd>& lb,
    const Eigen::Ref<const Eigen::VectorXd>& ub)
    : Constraint(v.rows(), num_vars, lb, ub) {
}
}
```
</blockquote></details>

I agree that this is a defect.

At a minimum, this method have an API disclaimer "Only to be called by subclass constructors." (and use Doxygen because its protected).

However in this case, I think an even easier solution than Hongkai's is to change ExpressionConstraint's call to the Constraint superclass constructor to pass
`num_var` as `v.sum().GetVariables().size()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[solvers/mathematical_program.h, line 1097 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7Ppm7Cu--9r4_B4OWB:-L7R-4Tkwh0TkNsa0Kvd:btlc4yc) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1097)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `prog.AddLinearConstraint` -> `prog.AddConstraint`.
</blockquote></details>

I think in the documentation, we should tell the user that our program will detect if the symbolic formula are linear, then they will be added as linear constraints, instead of generic nonlinear constraints.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 15 of 16 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[solvers/constraint.cc, line 177 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7Qjl-yt6_yKNhunICv:-L7Qjl-yt6_yKNhunICw:b-1cvdhb) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/constraint.cc#L177)):*
> ```cpp
>     const Eigen::Ref<const Eigen::VectorXd>& ub)
>     : Constraint(v.rows(), 0, lb, ub), expressions_(v) {
>   // Setup map_var_to_index and var_vec.
> ```

BTW `var_vec` seems like a stale typo for `vars_`.
(The punc is also a little weird in this line and the next.)

---

*[solvers/constraint.cc, line 195 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7QkYzndiSSGupWS7ai:-L7QkYzoztD_oA5FgZlu:b-5l4ek7) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/constraint.cc#L195)):*
> ```cpp
> 
> void ExpressionConstraint::DoEval(const Eigen::Ref<const Eigen::VectorXd>& x,
>             Eigen::VectorXd& y) const {
> ```

BTW Wrong indentation?

---

*[solvers/create_constraint.cc, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7Qna9GDblDHTksN2TR:-L7Qna9GDblDHTksN2TS:b-pfjpve) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/create_constraint.cc#L64)):*
> ```cpp
> 
>   if (is_equality_constraint) {
>     return ParseLinearEqualityConstraint(v, lb);
> ```

Unclear what the contract is for passing `ub` when `is_equality_constraint == true`.  I think we need better API docs and then enforcement here.  From the one single call site as of this PR, we have ub==lb==0.0 so either they must be identical, or even they must both be zero.  Either way, document in API and `DRAKE_DEMAND` here, I think?

---

*[solvers/integer_optimization_util.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R278i-vKZwjMrL7qc:-L7R278i-vKZwjMrL7qd:byq3yeq) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/integer_optimization_util.cc#L34)):*
> ```cpp
>                                 b1_xor_b2 <= 2 - b1 - b2 && 0 <= b1_xor_b2 &&
>                                 b1_xor_b2 <= 1));
> }
> ```

FYI No defect, because all of the conjunctions above were unreadable before and are still unreadable now, but consider formatting with a linebreak between each term in the conjunctions for readability.  Either clang-format off, or `// BR` at EOL to force a linebreak.

---

*[solvers/mathematical_program.cc, line 444 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R6-R_hg6wqqyJ1rqE:-L7R6-R_hg6wqqyJ1rqF:b-ym4cy3) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.cc#L444)):*
> ```cpp
>   Binding<Constraint> binding = internal::ParseConstraint(e, lb, ub);
>   Constraint* constraint = binding.evaluator().get();
>   if (dynamic_cast<LinearConstraint*>(constraint)) {
> ```

Ditto on the same downcast defect as the header file.  (Here, and the next two times it happens within this file in this PR.)

---

*[solvers/mathematical_program.cc, line 481 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R67v_VQp4EqDtzv9n:-L7R67v_VQp4EqDtzv9o:bgbhtn4) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.cc#L481)):*
> ```cpp
>     oss << "Formula " << f << " is non-linear.";
>     throw std::runtime_error(oss.str());
>   }
> ```

Rule-of-three code duplication to "check-for-linear-and-add-it-else-throw".  Since this is a `*.cc` file, it should be easy to add an anonymous helper?

---

*[solvers/mathematical_program.h, line 1286 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R42Rmt2Gl8CYbSwPw:-L7R42Rmt2Gl8CYbSwPx:bytt9vc) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1286)):*
> ```objc
>     Binding<Constraint> binding = internal::ParseConstraint(formulas);
>     Constraint* constraint = binding.evaluator().get();
>     if (dynamic_cast<LinearConstraint*>(constraint)) {
> ```

This fails GSG http://drake.mit.edu/styleguide/cppguide.html#Run-Time_Type_Information__RTTI_ because it is a speculative downcast.

N.B. the `internal::BindingDynamicCast` stuff is fine, because it's a downcast that is known ahead of time to succeed, because the (Drake-internal) calling code has formulated the parse in a way that will always be linear -- indeed, its an assertion failure if a `BindingDynamicCast` fails.

Here, its a user API and we don't know what will happen, so we don't know if the downcast will succeed or not.

To fix it, we need `bool Constraint::IsLinear() const` in the base class in order to do this properly.  Given that this is raising an exception, I'd be okay with a TODO and a next-week-or-two PR that fixes it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[solvers/evaluator_base.h, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7PwCQ5qsxfUCYBWF5i:-L7R7trARA59VN9NG5gI:behl5qc) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/evaluator_base.h#L131)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I agree that this is a defect.

At a minimum, this method have an API disclaimer "Only to be called by subclass constructors." (and use Doxygen because its protected).

However in this case, I think an even easier solution than Hongkai's is to change ExpressionConstraint's call to the Constraint superclass constructor to pass
`num_var` as `v.sum().GetVariables().size()`.
</blockquote></details>

I think passing `num_var` as `v.sum().GetVariables().size()` might be incorrect, for example if `v  << x(0), -x(0)`, then `v.sum() = 0`, and `num_var` would be 0, but actually it should be 1.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[solvers/evaluator_base.h, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7PwCQ5qsxfUCYBWF5i:-L7RBu2JeXYeXZZzqjPV:b-hhzylv) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/evaluator_base.h#L131)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think passing `num_var` as `v.sum().GetVariables().size()` might be incorrect, for example if `v  << x(0), -x(0)`, then `v.sum() = 0`, and `num_var` would be 0, but actually it should be 1.
</blockquote></details>

Aha, you're right!  We still do immediate simplification in operator+ (even though we shouldn't per #7904).

I guess we could do a hack like `v.array().abs().sum().GetVariables().size()` or something, but really a helper like `CountDistinctVariables(const Eigen::Ref<const VectorX<symbolic::Expression>>&)` would be best for clarity.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[solvers/constraint.cc, line 177 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7Qjl-yt6_yKNhunICv:-L7U0LMIcuQIjblErnVA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/constraint.cc#L177)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW `var_vec` seems like a stale typo for `vars_`.
(The punc is also a little weird in this line and the next.)
</blockquote></details>

Done.

---

*[solvers/constraint.cc, line 184 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7QTJwA2IZVuk2bBKJ9:-L7U6-wxaad_wFRSecuj:b-3xq8ye) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/constraint.cc#L184)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW `bazel test //solvers:constraint_test` does not fail when I comment out this line, though `//solvers:mathematical_program_test` does.  Do we want to have direct unit coverage here, or is indirect good enough?
</blockquote></details>

Method is removed.

---

*[solvers/constraint.cc, line 195 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7QkYzndiSSGupWS7ai:-L7U6A9f3Z6O0JPE_qHl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/constraint.cc#L195)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Wrong indentation?
</blockquote></details>

Done.

---

*[solvers/constraint.h, line 785 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7PgO_0zUpa994pRTiy:-L7U80hHbUY1hvcYmYto:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/constraint.h#L785)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, missing doxygen block.
</blockquote></details>

Done.

---

*[solvers/create_constraint.cc, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7Qna9GDblDHTksN2TR:-L7UAM8DJ-YIgaiq9j2s:b-52jl9t) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/create_constraint.cc#L64)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear what the contract is for passing `ub` when `is_equality_constraint == true`.  I think we need better API docs and then enforcement here.  From the one single call site as of this PR, we have ub==lb==0.0 so either they must be identical, or even they must both be zero.  Either way, document in API and `DRAKE_DEMAND` here, I think?
</blockquote></details>

Done.  I originally had it checking ub vs lb, but then wanted to avoid having tolerances... but as I started to write this DEMAND, I realized that having the default tolerance (0.) was the right solution.  So the strange extra API argument is gone.  

---

*[solvers/evaluator_base.h, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7PwCQ5qsxfUCYBWF5i:-L7U5lEPFGXRR-43v_o7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/evaluator_base.h#L131)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Aha, you're right!  We still do immediate simplification in operator+ (even though we shouldn't per #7904).

I guess we could do a hack like `v.array().abs().sum().GetVariables().size()` or something, but really a helper like `CountDistinctVariables(const Eigen::Ref<const VectorX<symbolic::Expression>>&)` would be best for clarity.
</blockquote></details>

Done.

---

*[solvers/integer_optimization_util.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R278i-vKZwjMrL7qc:-L7UB2EBXEeZm8Y-8GYP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/integer_optimization_util.cc#L34)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI No defect, because all of the conjunctions above were unreadable before and are still unreadable now, but consider formatting with a linebreak between each term in the conjunctions for readability.  Either clang-format off, or `// BR` at EOL to force a linebreak.
</blockquote></details>

Done.

---

*[solvers/mathematical_program.cc, line 444 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R6-R_hg6wqqyJ1rqE:-L7UE16yMUIGZOK6SmgH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.cc#L444)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ditto on the same downcast defect as the header file.  (Here, and the next two times it happens within this file in this PR.)
</blockquote></details>

Done.

---

*[solvers/mathematical_program.cc, line 481 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R67v_VQp4EqDtzv9n:-L7UDas74X8D7gTumxQ6:b6d7rcn) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.cc#L481)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Rule-of-three code duplication to "check-for-linear-and-add-it-else-throw".  Since this is a `*.cc` file, it should be easy to add an anonymous helper?
</blockquote></details>

not actually three.  the throw messages calls are subtly different.  i actually typed in the anonymous helper, but decided it was more gross (has to print Formula or Expression, etc) that the existing repeats.

---

*[solvers/mathematical_program.h, line 1046 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7PmSCGgAtzO3AMxdV0:-L7UECmtPojbQnJqGnJj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1046)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, please check the indentation.
</blockquote></details>

Done.

---

*[solvers/mathematical_program.h, line 1097 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7Ppm7Cu--9r4_B4OWB:-L7UFZeNQ5qa1phw_DH9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1097)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think in the documentation, we should tell the user that our program will detect if the symbolic formula are linear, then they will be added as linear constraints, instead of generic nonlinear constraints.
</blockquote></details>

Done.

---

*[solvers/mathematical_program.h, line 1286 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R42Rmt2Gl8CYbSwPw:-L7UFaJSoNNv_O-Ldz_2:b3ilbkt) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1286)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This fails GSG http://drake.mit.edu/styleguide/cppguide.html#Run-Time_Type_Information__RTTI_ because it is a speculative downcast.

N.B. the `internal::BindingDynamicCast` stuff is fine, because it's a downcast that is known ahead of time to succeed, because the (Drake-internal) calling code has formulated the parse in a way that will always be linear -- indeed, its an assertion failure if a `BindingDynamicCast` fails.

Here, its a user API and we don't know what will happen, so we don't know if the downcast will succeed or not.

To fix it, we need `bool Constraint::IsLinear() const` in the base class in order to do this properly.  Given that this is raising an exception, I'd be okay with a TODO and a next-week-or-two PR that fixes it.
</blockquote></details>

Done.  But I had to convince myself that is_linear_constraint was a special enough case to warrant calling it out like that.  Seems like a potential explosion of virtual methods.  (I'm still on the fence, but I've caved).  

I preferred is_linear_constraint to is_linear (my docs indicate why -- a generic constraint that happens to be linear should return false).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 14 of 14 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[common/symbolic_expression.cc, line 866 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7UNB9bJ-DyymO8Cf8q:-L7UNB9cM6tosUMimMyl:bt7uzrh) ([raw file](https://github.com/robotlocomotion/drake/blob/f0bc8d860a78556cc732ad5377d008e2f1854a2d/common/symbolic_expression.cc#L866)):*
> ```cpp
>       for (const Variable& var : v(i, j).GetVariables()) {
>         vars.insert(var);
>       }
> ```

FYI I think instead of the innermost for loop, doing
```c++
vars.Insert(v(i, j).GetVariables());
```
would be slightly cleaner.

---

*[common/symbolic_expression.h, line 903 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7UNpRVGUIFTXw-4mB1:-L7UNpRVGUIFTXw-4mB2:b-qrdy99) ([raw file](https://github.com/robotlocomotion/drake/blob/f0bc8d860a78556cc732ad5377d008e2f1854a2d/common/symbolic_expression.h#L903)):*
> ```objc
> 
> /// Counts the number of distinct variables in the matrix of expressions.
> int CountDistinctVariables(const Eigen::Ref<const MatrixX<Expression>>& v);
> ```

FYI You could imagine `Variables GetDistinctVariables` being slightly more general.  (Given NRVO, which is mostly happening already now and becomes mandatory in C++17, there wouldn't be an extra copy of the return value.)

---

*[solvers/mathematical_program.h, line 1286 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R42Rmt2Gl8CYbSwPw:-L7UP0HadwjsZFScCuLa:b-ojhvsv) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1286)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.  But I had to convince myself that is_linear_constraint was a special enough case to warrant calling it out like that.  Seems like a potential explosion of virtual methods.  (I'm still on the fence, but I've caved).  

I preferred is_linear_constraint to is_linear (my docs indicate why -- a generic constraint that happens to be linear should return false).
</blockquote></details>

OK Wow, well, looking at the code this ends up calling into (that I hadn't before), we have lots of RTTI checks (only some with TODOs) already within `mathematical_program.cc`.  Possibly being consistent with that badness and keeping this as a TODO instead of adding is_linear_constraint would have been be better, and then in the future we add some more palatable refactorings.  Up to you if you want to change it back.

I think what we'd want is that ParseConstraint returns a enum + Binding, where the enum says what kind of Binding we made.  (Or possibly the enum is buried in the Binding or Constraint, as some of the TODOs say.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-:-L7UR7PZG53l705zYBd3:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[common/symbolic_expression.cc, line 866 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7UNB9bJ-DyymO8Cf8q:-L7U_3Hn-R2eWXitaNGX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f0bc8d860a78556cc732ad5377d008e2f1854a2d/common/symbolic_expression.cc#L866)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I think instead of the innermost for loop, doing
```c++
vars.Insert(v(i, j).GetVariables());
```
would be slightly cleaner.
</blockquote></details>

Done.

---

*[common/symbolic_expression.h, line 903 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7UNpRVGUIFTXw-4mB1:-L7U_WOV_TKxg1N4Cokm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f0bc8d860a78556cc732ad5377d008e2f1854a2d/common/symbolic_expression.h#L903)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI You could imagine `Variables GetDistinctVariables` being slightly more general.  (Given NRVO, which is mostly happening already now and becomes mandatory in C++17, there wouldn't be an extra copy of the return value.)
</blockquote></details>

Done.

---

*[solvers/mathematical_program.h, line 1286 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R42Rmt2Gl8CYbSwPw:-L7Ua5jDXWejU028iTxO:b2gjdu7) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1286)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

OK Wow, well, looking at the code this ends up calling into (that I hadn't before), we have lots of RTTI checks (only some with TODOs) already within `mathematical_program.cc`.  Possibly being consistent with that badness and keeping this as a TODO instead of adding is_linear_constraint would have been be better, and then in the future we add some more palatable refactorings.  Up to you if you want to change it back.

I think what we'd want is that ParseConstraint returns a enum + Binding, where the enum says what kind of Binding we made.  (Or possibly the enum is buried in the Binding or Constraint, as some of the TODOs say.)
</blockquote></details>

reverted.  i really dislike putting in a method only to declare (in a redundant, and potentially fragile via bit rot way) the information already contained in the inheritance hierarchy. 
sounds like this is not a TODO for these lines, but rather a larger issue: https://github.com/RobotLocomotion/drake/issues/8349

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[common/symbolic_expression.h, line 903 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7UNpRVGUIFTXw-4mB1:-L7UdndiT72uwy9qjID1:bxvqbn2) ([raw file](https://github.com/robotlocomotion/drake/blob/f0bc8d860a78556cc732ad5377d008e2f1854a2d/common/symbolic_expression.h#L903)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

OK The API comment is slightly stale now.

---

*[solvers/mathematical_program.h, line 1286 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7R42Rmt2Gl8CYbSwPw:-L7UdrRJrs1h-5RwvDPr:b-8ms9ix) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1286)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

reverted.  i really dislike putting in a method only to declare (in a redundant, and potentially fragile via bit rot way) the information already contained in the inheritance hierarchy. 
sounds like this is not a TODO for these lines, but rather a larger issue: https://github.com/RobotLocomotion/drake/issues/8349
</blockquote></details>

OK Yeah, it will require some more thought.  GSG also says not to re-implement RTTI, which the `is_foo` method(s) would kinda be.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 17 of 18 files reviewed at latest revision, 3 unresolved discussions.

---

*[common/symbolic_expression.h, line 903 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7UNpRVGUIFTXw-4mB1:-L7UgWZEyMZlSOwF49PU:bkz5ckv) ([raw file](https://github.com/robotlocomotion/drake/blob/f0bc8d860a78556cc732ad5377d008e2f1854a2d/common/symbolic_expression.h#L903)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

OK The API comment is slightly stale now.
</blockquote></details>

Done.  Good catch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 6 of 14 files at r2, 6 of 7 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[solvers/mathematical_program.h, line 1097 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8338#-L7Ppm7Cu--9r4_B4OWB:-L7Uq1hgDOH1UgN555UF:b-n1hudd) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1097)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Not done? I think we should also mention that we will not parse a symbolic formula to a (rotated) Lorentz cone constraint, or a PSD constraint. For example, if we pass the symbolic formulat
```
x(0) >= sqrt(x(1)**2 + x(2)**2)
```
This will be added as a generic nonlinear constraint, rather than a Lorentz cone constraint.
</blockquote></details>

I added the details to the overload above that takes a single formula.  I've added the overload tag here to reference that.  I really dislike copying these details everywhere, since it will be prone to errors if/when we change the implementation.  

Also, I think that the description (which says the types of constraints that can result) is better as a white list than a black list, which is definitely going to get out of date.  

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[solvers/mathematical_program.h, line 1097 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8338#-L7Ppm7Cu--9r4_B4OWB:-L7UwMzKhm2KfHZBKjJc:b-wsvasg) ([raw file](https://github.com/robotlocomotion/drake/blob/494b9df782be7578ae00ad06fdcd1e465b13d3e6/solvers/mathematical_program.h#L1097)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

I added the details to the overload above that takes a single formula.  I've added the overload tag here to reference that.  I really dislike copying these details everywhere, since it will be prone to errors if/when we change the implementation.  

Also, I think that the description (which says the types of constraints that can result) is better as a white list than a black list, which is definitely going to get out of date.  

</blockquote></details>

Thanks for the explanation!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8338)*
<!-- Sent from Reviewable.io -->
